### DISCUSSION TITLE
Mironaut 4 migration issue: controller NOT_FOUND

### DISCUSSION THREAD

**Discussion by ethanhere** [micronaut-projects/micronaut-core#10389] [General]
As I migrate our current application to micronaut 4, the following code was working with micronaut 3 but got into NOT_FOUND with micronaut 4. It's really puzzling:
Controller:
@Context
@Controller("/")
@ExecuteOn(TaskExecutors.BLOCKING)
@Validated
@Requires(env = "test")
public class TestController {

    @Inject
    public TestController () {
    }

    @Get(uri = "/endpoint")
    @Secured(SecurityRule.IS_AUTHENTICATED)
    public HttpResponse<TestResponse> endpoint(String authorization) {
…
 
Client:
@Client("/")
@Requires(env = "test")
public interface TestClient {

    @Get("/endpoint")
    public HttpResponse<TestResponse> endpoint(@Header String authorization);
..
 
Test:
@MicronautTest(environments = "binding-test")
public class BindingAppTest {

    @Inject
    private TestClient client;

    @Test
    public void testEndpointMetadataBinding() {

        String authHeader;
        final HttpResponse< TestResponse > response = this.client. endpoint (authHeader);
        Assertions.assertEquals(HttpStatus.OK, response.getStatus());
    }
 
The test expects to get OK but actually got NOT_FOUND. The log shows:
15:43:06.753 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.http.server.netty.NettyRequestLifecycle - Request GET /endpoint
15:43:06.754 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.http.server.RequestLifecycle - No matching route: GET /endpoint

What makes the controller not matched with route in micronaut 4?

**Comment by ethanhere** [comment]
I also found the following log that may be helpful:
11:51:19.593 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.DefaultBeanContext - Finding candidate beans for type: EndpointConfiguration E
11:51:19.593 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.DefaultBeanContext - No bean candidates found for type: EndpointConfiguration E
11:51:19.593 [default-nioEventLoopGroup-1-3] TRACE io.micronaut.context.DefaultBeanContext - Looking up existing bean for key: defaultConfiguration
11:51:19.593 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.DefaultBeanContext - Finding candidate beans for type: EndpointDefaultConfiguration defaultConfiguration
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.DefaultBeanContext -   class io.micronaut.management.endpoint.EndpointDefaultConfiguration null Definition: io.micronaut.management.endpoint.EndpointDefaultConfiguration
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.DefaultBeanContext - Found concrete candidate [Definition: io.micronaut.management.endpoint.EndpointDefaultConfiguration] for type: defaultConfiguration 
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.DefaultBeanContext - Finding candidate beans for qualifier: @Endpoint
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.loggers.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.condition.Condition - Bean [Definition: io.micronaut.management.endpoint.loggers.LoggersEndpoint] will not be loaded due to failing conditions:
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.condition.Condition - * Custom condition [class io.micronaut.management.endpoint.EndpointEnabledCondition] failed evaluation
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.stop.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.condition.Condition - Bean [Definition: io.micronaut.management.endpoint.stop.ServerStopEndpoint] will not be loaded due to failing conditions:
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.condition.Condition - * Custom condition [class io.micronaut.management.endpoint.EndpointEnabledCondition] failed evaluation
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.env.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.condition.Condition - Bean [Definition: io.micronaut.management.endpoint.env.EnvironmentEndpoint] will not be loaded due to failing conditions:
11:51:19.594 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.condition.Condition - * Custom condition [class io.micronaut.management.endpoint.EndpointEnabledCondition] failed evaluation
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.beans.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.refresh.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.threaddump.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.594 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.info.enabled
11:51:19.595 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.595 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.routes.enabled
11:51:19.595 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.595 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.health.enabled
11:51:19.595 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.all.enabled
11:51:19.595 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.lifecycle - Created bean [io.micronaut.management.endpoint.EndpointSensitivityProcessor@1ca3b9b3] from definition [Definition: io.micronaut.management.endpoint.EndpointSensitivityProcessor] with qualifier [null]
11:51:19.595 [default-nioEventLoopGroup-1-3] DEBUG io.micronaut.context.lifecycle - Created bean [io.micronaut.security.rules.SensitiveEndpointRule@4be3556d] from definition [Definition: io.micronaut.security.rules.SensitiveEndpointRule] with qualifier [null]

Especially 11:51:19.595 [default-nioEventLoopGroup-1-3] TRACE i.m.context.DefaultApplicationContext$RuntimeConfiguredEnvironment - No value found for property: endpoints.routes.enabled

I guess something missing to enable endpoints route. I have no idea what cause this and how to correct it.

**Reply by Jonathan-Zollinger** [reply]
I wonder if it's a [breaking change from micronaut security](https://micronaut-projects.github.io/micronaut-security/2.0.0.RC1/guide/index.html#breaks). 

For the sake of troubleshooting, what if you disable endpoint security, which I believe would just be `endpoints.routes.sensitive=false`?

**Comment by sdelamo** [comment]
do you have a sample app?
